### ISSUE TITLE
Can only test build environment on Linux

### ISSUE THREAD

**Issue by XAMPPRocky** [rust-lang/docs.rs#522]
The [current instructions in the wiki](https://github.com/rust-lang/docs.rs/wiki/Making-changes-to-the-build-environment) for building and testing additonal packages to the build environment do not work on macOS for two distinct reasons.

* `realpath` is not available on macOS by default. This is fixable with `brew install coreutils`
* The current instructions mount your rust installation and using that when building in Ubuntu, which does not work if you're on macOS.

### Potential solutions
* Provide a OS (At least *nix) agnostic way to test build environment.
* Add seperate instructions for macOS.

**Comment by jyn514** [comment]
Do you mind volunteering your time to test this solution?

```
realpath() {
    (cd "$1" && pwd -P)
}
rustup install nightly-x86_64-unknown-linux-gnu
rustup_home="$(rustup show home)"
cargo_home="$([ "$CARGO_HOME" = "" ] && echo ~/.cargo || echo "$CARGO_HOME")"
cd /path/to/your/crate
CONTAINER="$("docker" "create" \
    "-v" "$(realpath ./target):/opt/rustwide/target:rw,Z" \
    "-v" "$(realpath .):/opt/rustwide/workdir:ro,Z" \
    "-v" "$cargo_home:/opt/rustwide/cargo-home:ro,Z" \
    "-v" "$rustup_home:/opt/rustwide/rustup-home:ro,Z" \
    "-e" "SOURCE_DIR=/opt/rustwide/workdir" "-e" "MAP_USER_ID=$(id -u)" \
    "-e" "CARGO_TARGET_DIR=/opt/rustwide/target" "-e" "RUSTFLAGS=" \
    "-e" 'RUSTDOCFLAGS=-Z unstable-options 
        --resource-suffix -20191116-1.41.0-nightly-5c5b8afd8 
        --static-root-path / --disable-per-crate-search' \
    "-e" "CARGO_HOME=/opt/rustwide/cargo-home" \
    "-e" "RUSTUP_HOME=/opt/rustwide/rustup-home" \
    "-w" "/opt/rustwide/workdir" "-m" "3221225472" "--network" "none" \
    "build-env" "/opt/rustwide/cargo-home/bin/cargo" \
        "+nightly" "doc" --offline --locked "--lib" "--no-deps" "--target" "x86_64-unknown-linux-gnu")"
docker start -ai "$CONTAINER"
```

**Comment by XAMPPRocky** [comment]
These did not work, I ran the below in the directory for the crate and received the following output.

```bash
realpath() {
    (cd "$1" && pwd -P)
}
rustup install nightly-x86_64-unknown-linux-gnu
rustup_home="$(rustup show home)"
cargo_home="$([ "$CARGO_HOME" = "" ] && echo ~/.cargo || echo "$CARGO_HOME")"
CONTAINER="$("docker" "create" \
    "-v" "$(realpath ./target):/opt/rustwide/target:rw,Z" \
    "-v" "$(realpath .):/opt/rustwide/workdir:ro,Z" \
    "-v" "$cargo_home:/opt/rustwide/cargo-home:ro,Z" \
    "-v" "$rustup_home:/opt/rustwide/rustup-home:ro,Z" \
    "-e" "SOURCE_DIR=/opt/rustwide/workdir" "-e" "MAP_USER_ID=$(id -u)" \
    "-e" "CARGO_TARGET_DIR=/opt/rustwide/target" "-e" "RUSTFLAGS=" \
    "-e" 'RUSTDOCFLAGS=-Z unstable-options
        --resource-suffix -20191116-1.41.0-nightly-5c5b8afd8
        --static-root-path / --disable-per-crate-search' \
    "-e" "CARGO_HOME=/opt/rustwide/cargo-home" \
    "-e" "RUSTUP_HOME=/opt/rustwide/rustup-home" \
    "-w" "/opt/rustwide/workdir" "-m" "3221225472" "--network" "none" \
    "build-env" "/opt/rustwide/cargo-home/bin/cargo" \
        "+nightly" "doc" --offline --locked "--lib" "--no-deps" "--target" "x86_64-unknown-linux-gnu")"
docker start -ai "$CONTAINER"
info: syncing channel updates for 'nightly-x86_64-unknown-linux-gnu'

  nightly-x86_64-unknown-linux-gnu unchanged - (error reading rustc version)

info: checking for self-updates
/opt/rustwide/cargo-home/bin/cargo: 1: /opt/rustwide/cargo-home/bin/cargo: �����8: not found
/opt/rustwide/cargo-home/bin/cargo: 2: /opt/rustwide/cargo-home/bin/cargo: ��H__PAGEZERO�__TEXTPVP__text__TEXT���F��__stubs__TEXT: not found
/opt/rustwide/cargo-home/bin/cargo: 3: /opt/rustwide/cargo-home/bin/cargo: Syntax error: "(" unexpected
```

**Comment by jyn514** [comment]
Ah right, it's still trying to use the default version in `bin/cargo`. One last try?

```
realpath() {
    (cd "$1" && pwd -P)
}
rustup install nightly-x86_64-unknown-linux-gnu
rustup_home="$(rustup show home)"
cargo_home="$([ "$CARGO_HOME" = "" ] && echo ~/.cargo || echo "$CARGO_HOME")"
CONTAINER="$("docker" "create" \
    "-v" "$(realpath ./target):/opt/rustwide/target:rw,Z" \
    "-v" "$(realpath .):/opt/rustwide/workdir:ro,Z" \
    "-v" "$cargo_home:/opt/rustwide/cargo-home:ro,Z" \
    "-v" "$rustup_home:/opt/rustwide/rustup-home:ro,Z" \
    "-e" "SOURCE_DIR=/opt/rustwide/workdir" "-e" "MAP_USER_ID=$(id -u)" \
    "-e" "CARGO_TARGET_DIR=/opt/rustwide/target" "-e" "RUSTFLAGS=" \
    "-e" 'RUSTDOCFLAGS=-Z unstable-options
        --resource-suffix -20191116-1.41.0-nightly-5c5b8afd8
        --static-root-path / --disable-per-crate-search' \
    "-e" "CARGO_HOME=/opt/rustwide/cargo-home" \
    "-e" "RUSTUP_HOME=/opt/rustwide/rustup-home" \
    "-w" "/opt/rustwide/workdir" "-m" "3221225472" "--network" "none" \
    "rustops/crates-build-env" bash -c 'export PATH="$PATH:/opt/rustwide/rustup-home/toolchains/nightly-x86_64-unknown-linux-gnu/bin" && 
            "cargo" "doc" --offline --locked "--lib" "--no-deps" "--target" "x86_64-unknown-linux-gnu"')"
docker start -ai "$CONTAINER"
```

If this doesn't work we probably won't be able to use the host toolchain at all and I'll have to set up _another_ intermediary docker container :(

**Comment by jyn514** [comment]
As a workaround, you can install rust in the build container directly: https://blog.rust-lang.org/2019/09/18/upcoming-docsrs-changes.html#how-to-prepare-for-the-changes

**Comment by jyn514** [comment]
Ok, clearly the docker container isn't working for people outside the very narrow case of x86_64-unknown-linux-gnu. I think a better long-term solution is to allow the docker image for docs.rs to be configured.

Mentoring instructions:

- In [`RustwideBuilder::init`](https://github.com/rust-lang/docs.rs/blob/master/src/docbuilder/rustwide_builder.rs#L80), check for an environment variable (maybe `DOCS_RS_LOCAL_IMAGE`)
- If it is set,
    - Create a `SandboxImage` using [`SandboxImage::local`](https://docs.rs/rustwide/0.3.2/rustwide/cmd/struct.SandboxImage.html)
    - In the same spot, call [`.sandbox_image()`](https://docs.rs/rustwide/0.3.2/rustwide/struct.WorkspaceBuilder.html#method.sandbox_image) on the `WorkspaceBuilder` with the image as an argument.
- Update [the README](https://github.com/rust-lang/docs.rs/#changing-the-build-environment) to show how to use a custom image (i.e. `DOCS_RS_LOCAL_IMAGE=build-env cargo run`)

**Comment by jyn514** [comment]
I'm going to work on this.

**Comment by jyn514** [comment]
Fixed in #593
